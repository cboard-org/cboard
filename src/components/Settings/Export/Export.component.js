import React from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';
import Button from '@material-ui/core/Button';
import Menu from '@material-ui/core/Menu';
import MenuItem from '@material-ui/core/MenuItem';
import Paper from '@material-ui/core/Paper';
import List from '@material-ui/core/List';
import CircularProgress from '@material-ui/core/CircularProgress';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import ListItemSecondaryAction from '@material-ui/core/ListItemSecondaryAction';
import FullScreenDialog from '../../UI/FullScreenDialog';
import messages from './Export.messages';

import './Export.css';

const propTypes = {
  /**
   * Callback fired when clicking the export Cboard button
   */
  onExportClick: PropTypes.func.isRequired,
  /**
   * Callback fired when clicking the back button
   */
  onClose: PropTypes.func
};

class Export extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      exportMenu: null,
      loading: false
    };
  }

  openMenu(e) {
    this.setState({ exportMenu: e.currentTarget });
  }

  closeMenu() {
    this.setState({ exportMenu: null });
  }

  onExportClick(type = 'cboard') {
    const doneCallback = () => {
      this.setState({ loading: false });
    };

    this.setState({ loading: true, exportMenu: null }, () => {
      this.props.onExportClick(type, doneCallback);
    });
  }

  render() {
    const { onClose } = this.props;
    return (
      <div className="Export">
        <FullScreenDialog
          open
          title={<FormattedMessage {...messages.export} />}
          onClose={onClose}
        >
          <Paper>
            <List>
              <ListItem>
                <ListItemText
                  primary={<FormattedMessage {...messages.export} />}
                  secondary={
                    <FormattedMessage
                      {...messages.exportSecondary}
                      values={{
                        link: (
                          <a href="http://www.openboardformat.org/">
                            OpenBoard
                          </a>
                        )
                      }}
                    />
                  }
                />
                <ListItemSecondaryAction>
                  <div className="Export__ButtonContainer">
                    {this.state.loading && (
                      <CircularProgress
                        size={25}
                        className="Export__ButtonContainer--spinner"
                        thickness={7}
                      />
                    )}
                    <Button
                      id="export-button"
                      disabled={this.state.loading}
                      onClick={this.openMenu.bind(this)}
                    >
                      <FormattedMessage {...messages.export} />
                    </Button>
                    <Menu
                      id="export-menu"
                      anchorEl={this.state.exportMenu}
                      open={Boolean(this.state.exportMenu)}
                      onClose={this.closeMenu.bind(this)}
                    >
                      <MenuItem
                        onClick={this.onExportClick.bind(this, 'cboard')}
                      >
                        Cboard
                      </MenuItem>
                      <MenuItem
                        onClick={this.onExportClick.bind(this, 'openboard')}
                      >
                        OpenBoard
                      </MenuItem>
                      <MenuItem onClick={this.onExportClick.bind(this, 'pdf')}>
                        PDF
                      </MenuItem>
                      {/*
                      <MenuItem onClick={this.onExportClick.bind(this, 'image')}>
                        Image
                      </MenuItem>
                      */}
                    </Menu>
                  </div>
                </ListItemSecondaryAction>
              </ListItem>
            </List>
          </Paper>
        </FullScreenDialog>
      </div>
    );
  }
}

Export.propTypes = propTypes;

export default Export;
