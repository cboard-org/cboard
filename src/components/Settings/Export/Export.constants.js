export const CBOARD_OBF_CONSTANTS = {
  DATA_URL: 'http://myboards.net/api/v1/boards/',
  URL: 'http://myboards.net/boards/',
  LICENSE: {
    type: 'CC-By',
    copyright_notice_url: 'http://creativecommons.org/licenses/by',
    source_url: 'https://github.com/shayc/cboard',
    author_name: 'Shay Cojocaru',
    author_url: 'https://github.com/shayc',
    author_email: 'shayc@outlook.com'
  }
};

export const CBOARD_ZIP_OPTIONS = {
  type: 'blob',
  compression: 'DEFLATE',
  platform: 'UNIX'
};

export const CBOARD_COLUMNS = 6;
export const CBOARD_EXT_PREFIX = 'ext_cboard_';
export const CBOARD_EXT_PROPERTIES = ['labelKey', 'nameKey'];

export const EXPORT_CONFIG_BY_TYPE = {
  cboard: {
    filename: 'board.json',
    callback: 'cboardExportAdapter'
  },
  openboard: {
    filename: 'board.obz',
    callback: 'openboardExportAdapter'
  },
  pdf: {
    filename: 'board.pdf',
    callback: 'pdfExportAdapter'
  }
};
