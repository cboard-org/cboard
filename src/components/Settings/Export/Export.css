.Export__ButtonContainer {
  position: relative;
}

.Export__ButtonContainer--spinner {
  vertical-align: middle;
}
