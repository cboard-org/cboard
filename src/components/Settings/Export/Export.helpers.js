import JSZip from 'jszip';
import axios from 'axios';
import pdfMake from 'pdfmake/build/pdfmake';
import pdfFonts from 'pdfmake/build/vfs_fonts';
import { saveAs } from 'file-saver';
import {
  EXPORT_CONFIG_BY_TYPE,
  CBOARD_OBF_CONSTANTS,
  CBOARD_COLUMNS,
  CBOARD_EXT_PREFIX,
  CBOARD_EXT_PROPERTIES,
  CBOARD_ZIP_OPTIONS
} from './Export.constants';

pdfMake.vfs = pdfFonts.pdfMake.vfs;

const imageElement = new Image();

function toSnakeCase(str) {
  const value = str.replace(/([A-Z])/g, $1 => '_' + $1.toLowerCase());
  return value.startsWith('_') ? value.slice(1) : value;
}

function getOBFButtonProps(tile = {}, intl) {
  const button = {};

  const tileExtProps = CBOARD_EXT_PROPERTIES.filter(key => !!tile[key]);
  tileExtProps.forEach(key => {
    const keyWithPrefix = `${CBOARD_EXT_PREFIX}${toSnakeCase(key)}`;
    button[keyWithPrefix] = tile[key];
  });

  const label = tile.label || tile.labelKey || '';
  button.label = label.length ? intl.formatMessage({ id: label }) : label;

  if (tile.action) {
    button.action = tile.action;
  }

  if (tile.vocalization) {
    button.vocalization = tile.vocalization;
  }

  if (tile.borderColor) {
    button['border_color'] = tile.borderColor;
  }

  if (tile.backgroundColor) {
    button['background_color'] = tile.backgroundColor;
  }

  return button;
}

function getBase64Image(base64Str = '') {
  const [prefix, base64Data] = base64Str.split(',');
  const contentType = prefix.split(':')[1].split(';')[0];
  const byteString = atob(base64Data);

  // https://gist.github.com/fupslot/5015897
  const ab = new ArrayBuffer(byteString.length);
  const ia = new Uint8Array(ab);
  for (var i = 0; i < byteString.length; i++) {
    ia[i] = byteString.charCodeAt(i);
  }

  return {
    data: ab,
    content_type: contentType
  };
}

async function boardToOBF(boardsMap, board = {}, intl) {
  if (!board.tiles || board.tiles.length < 1) {
    return { obf: null, images: null };
  }

  const images = {};
  const fetchedImages = {};
  const grid = new Array(Math.ceil(board.tiles.length / CBOARD_COLUMNS));
  let currentRow = 0;
  const buttons = await Promise.all(
    board.tiles.map(async (tile, i) => {
      currentRow =
        i >= (currentRow + 1) * CBOARD_COLUMNS ? currentRow + 1 : currentRow;

      if (grid[currentRow]) {
        grid[currentRow].push(tile.id);
      } else {
        grid[currentRow] = [tile.id];
      }

      const button = {
        id: tile.id,
        ...getOBFButtonProps(tile, intl)
      };

      if (tile.image && tile.image.length) {
        let imageResponse = null;
        let url = '';
        let contentType = '';
        let fetchedImageID = `custom/${board.name ||
          board.nameKey}/${tile.label || tile.labelKey || tile.id}`;

        if (tile.image.startsWith('data:')) {
          imageResponse = getBase64Image(tile.image);
          contentType = imageResponse['content_type'];
          const defaultExtension =
            contentType.indexOf('/') >= 0 ? contentType.split('/')[1] : '';
          fetchedImageID = defaultExtension.length
            ? `${fetchedImageID}.${defaultExtension}`
            : fetchedImageID;
          url = `/${fetchedImageID}`;
        } else {
          url = tile.image.startsWith('/') ? tile.image : `/${tile.image}`;
          fetchedImageID = tile.image;
          try {
            imageResponse = await axios({
              method: 'get',
              url,
              responseType: 'arraybuffer'
            });

            contentType = imageResponse.headers['content-type'];
          } catch (e) {}
        }

        if (imageResponse) {
          const imageID = `${board.id}_${tile.image}`;
          fetchedImages[fetchedImageID] = imageResponse;
          button['image_id'] = imageID;
          images[imageID] = {
            id: imageID,
            path: `images${url}`,
            content_type: contentType,
            width: 300,
            height: 300
          };
        }
      }

      if (tile.loadBoard && boardsMap[tile.loadBoard]) {
        const loadBoardData = boardsMap[tile.loadBoard];
        button['load_board'] = {
          name: loadBoardData.nameKey
            ? intl.formatMessage({ id: loadBoardData.nameKey })
            : '',
          path: `boards/${tile.loadBoard}.obf`
        };
      }

      return button;
    })
  );

  const lastGridRowDiff = CBOARD_COLUMNS - grid[grid.length - 1].length;
  if (lastGridRowDiff > 0) {
    const emptyButtons = new Array(lastGridRowDiff).map(() => null);
    grid[grid.length - 1] = grid[grid.length - 1].concat(emptyButtons);
  }

  const obf = {
    format: 'open-board-0.1',
    id: board.id,
    locale: intl.locale,
    name: intl.formatMessage({ id: board.nameKey || board.id }),
    url: `${CBOARD_OBF_CONSTANTS.URL}${board.id}`,
    license: CBOARD_OBF_CONSTANTS.LICENSE,
    images: Object.values(images),
    buttons,
    sounds: [],
    grid: {
      rows: grid.length,
      columns: CBOARD_COLUMNS,
      order: grid
    },
    description_html: board.nameKey
      ? intl.formatMessage({ id: board.nameKey })
      : ''
  };

  const boardExtProps = CBOARD_EXT_PROPERTIES.filter(key => !!board[key]);
  boardExtProps.forEach(key => {
    const keyWithPrefix = `${CBOARD_EXT_PREFIX}${toSnakeCase(key)}`;
    obf[keyWithPrefix] = board[key];
  });

  return { obf, images: fetchedImages };
}

function getPDFTileData(tile, intl) {
  const label = tile.label || tile.labelKey || '';
  return {
    label: label.length ? intl.formatMessage({ id: label }) : label,
    image: tile.image || ''
  };
}

async function toDataURL(url, styles = {}, outputFormat = 'image/jpeg') {
  return new Promise(resolve => {
    imageElement.crossOrigin = 'Anonymous';
    imageElement.onload = function() {
      const canvas = document.createElement('CANVAS');
      const ctx = canvas.getContext('2d');
      const backgroundColor = styles.backgroundColor || 'white';
      const borderColor = styles.borderColor || null;
      canvas.height = 150;
      canvas.width = 150;

      ctx.fillStyle = backgroundColor;
      ctx.fillRect(0, 0, canvas.width, canvas.height);

      if (borderColor) {
        ctx.strokeStyle = borderColor;
        ctx.lineWidth = 2;
        ctx.strokeRect(0, 0, canvas.width, canvas.height);
      }

      let widthFix = 1;
      let heightFix = 1;
      const needToScale = this.naturalWidth > 150 || this.naturalHeight > 150;
      if (needToScale) {
        widthFix = 150 / this.naturalWidth;
        heightFix = 150 / this.naturalHeight;
      }

      ctx.drawImage(
        this,
        0,
        0,
        this.naturalWidth * widthFix,
        this.naturalHeight * heightFix
      );

      if (borderColor) {
        ctx.strokeStyle = borderColor;
        ctx.lineWidth = 3;
        ctx.strokeRect(0, 0, 150, 150);
      }

      const dataURL = canvas.toDataURL(outputFormat);
      resolve(dataURL);
    };

    imageElement.src = url;
    if (imageElement.complete || imageElement.complete === undefined) {
      imageElement.src = url;
    }
  });
}

async function generatePDFBoard(board, intl, breakPage = true) {
  const header = {
    text: intl.formatMessage({ id: board.nameKey || board.id })
  };

  const table = {
    table: {
      widths: '16%',
      body: [{}]
    },
    layout: 'noBorders'
  };

  if (breakPage) {
    table.pageBreak = 'after';
  }

  if (!board.tiles || !board.tiles.length) {
    return [header, table];
  }

  // Do a grid with 2n rows
  const grid = new Array(Math.ceil(board.tiles.length / CBOARD_COLUMNS) * 2);
  let currentRow = 0;

  await board.tiles.reduce(async (prev, tile, i) => {
    // Wait for previous tile
    await prev;

    const { label, image } = getPDFTileData(tile, intl);
    currentRow =
      i >= (currentRow + 1) * CBOARD_COLUMNS ? currentRow + 1 : currentRow;
    const fixedRow = currentRow * 2;
    let imageData = '';

    if (image && image.length) {
      let dataURL = image;
      if (
        !image.startsWith('data:') ||
        image.startsWith('data:image/svg+xml')
      ) {
        let url = image;

        const styles = {};

        if (tile.backgroundColor) {
          styles.backgroundColor = tile.backgroundColor;
        }
        if (tile.borderColor) {
          styles.borderColor = tile.borderColor;
        }

        dataURL = await toDataURL(url, styles);
      }

      imageData = {
        image: dataURL,
        alignment: 'center',
        width: '100'
      };
    }

    const labelData = {
      text: label,
      alignment: 'center'
    };

    if (grid[fixedRow]) {
      grid[fixedRow].push(labelData);
      grid[fixedRow + 1].push(imageData);
    } else {
      grid[fixedRow] = [labelData];
      grid[fixedRow + 1] = [imageData];
    }

    return grid;
  }, Promise.resolve());

  const lastGridRowDiff = CBOARD_COLUMNS - grid[grid.length - 2].length; // labels row
  if (lastGridRowDiff > 0) {
    const emptyCells = new Array(lastGridRowDiff).fill('');
    grid[grid.length - 2] = grid[grid.length - 2].concat(emptyCells); // labels
    grid[grid.length - 1] = grid[grid.length - 1].concat(emptyCells); // images
  }

  table.table.body = grid;

  return [header, table];
}

export async function openboardExportAdapter(boards = [], intl) {
  const boardsLength = boards.length;
  const boardsForManifest = {};
  const imagesMap = {};
  const zip = new JSZip();

  const boardsMap = boards.reduce((prev, current) => {
    prev[current.id] = current;
    return prev;
  }, {});

  for (let i = 0; i < boardsLength; i++) {
    const board = boards[i];
    const boardMapFilename = `boards/${board.id}.obf`;
    const { obf, images } = await boardToOBF(boardsMap, board, intl);

    if (!obf) {
      continue;
    }

    zip.file(boardMapFilename, JSON.stringify(obf));

    const imagesKeys = Object.keys(images);
    imagesKeys.forEach(key => {
      const imageFilename = `images/${key}`;
      zip.file(imageFilename, images[key].data);
      imagesMap[key] = imageFilename;
    });

    boardsForManifest[board.id] = boardMapFilename;
  }

  const root = boardsForManifest.root
    ? boardsForManifest.root
    : boardsForManifest[Object.keys(boardsMap)[0]];

  const manifest = {
    format: 'open-board-0.1',
    root,
    paths: {
      boards: boardsForManifest,
      images: imagesMap
    }
  };

  zip.file('manifest.json', JSON.stringify(manifest));

  zip.generateAsync(CBOARD_ZIP_OPTIONS).then(content => {
    saveAs(content, EXPORT_CONFIG_BY_TYPE.openboard.filename);
  });
}

export async function cboardExportAdapter(boards = []) {
  const jsonData = new Blob([JSON.stringify(boards)], {
    type: 'text/json;charset=utf-8;'
  });

  // IE11 & Edge
  if (navigator.msSaveBlob) {
    navigator.msSaveBlob(jsonData, EXPORT_CONFIG_BY_TYPE.cboard.filename);
  } else {
    // In FF link must be added to DOM to be clicked
    const link = document.createElement('a');
    link.href = window.URL.createObjectURL(jsonData);
    link.setAttribute('download', EXPORT_CONFIG_BY_TYPE.cboard.filename);
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
  }
}

export async function pdfExportAdapter(boards = [], intl) {
  const docDefinition = {
    pageSize: 'A4',
    pageOrientation: 'landscape',
    content: []
  };
  const lastBoardIndex = boards.length - 1;
  const content = await boards.reduce(async (prev, board, i) => {
    const prevContent = await prev;
    const breakPage = i !== lastBoardIndex;
    const boardPDFData = await generatePDFBoard(board, intl, breakPage);
    return prevContent.concat(boardPDFData);
  }, Promise.resolve([]));

  docDefinition.content = content;

  pdfMake.createPdf(docDefinition).download(EXPORT_CONFIG_BY_TYPE.pdf.filename);
}

export default {
  openboardExportAdapter,
  cboardExportAdapter,
  pdfExportAdapter
};
