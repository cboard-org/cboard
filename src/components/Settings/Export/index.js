export { default } from './Export.container';
