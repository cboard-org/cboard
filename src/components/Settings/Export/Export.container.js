import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { injectIntl, intlShape } from 'react-intl';

import { showNotification } from '../../Notifications/Notifications.actions';
import Export from './Export.component';
import { EXPORT_CONFIG_BY_TYPE } from './Export.constants';

export class ExportContainer extends PureComponent {
  static propTypes = {
    boards: PropTypes.array.isRequired,
    history: PropTypes.object.isRequired,
    intl: intlShape.isRequired
  };

  handleExportClick = async (type = 'cboard', doneCallback) => {
    const exportConfig = EXPORT_CONFIG_BY_TYPE[type];
    const EXPORT_HELPERS = await import('./Export.helpers');

    if (
      !exportConfig ||
      !exportConfig.callback ||
      !EXPORT_HELPERS[exportConfig.callback]
    ) {
      return false;
    }

    const { boards, intl } = this.props;

    await EXPORT_HELPERS[exportConfig.callback](boards, intl);

    doneCallback();
  };

  openboardExportAdapter(boards) {
    return boards;
  }

  render() {
    const { boards, history } = this.props;

    return (
      <Export
        boards={boards}
        onExportClick={this.handleExportClick}
        onClose={history.goBack}
      />
    );
  }
}

const mapStateToProps = state => ({
  boards: state.board.boards
});

const mapDispatchToProps = {
  showNotification
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(injectIntl(ExportContainer));
