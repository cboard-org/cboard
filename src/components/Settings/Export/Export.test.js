import React from 'react';
import { shallowMatchSnapshot } from '../../../common/test_utils';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Export from './Export.component';

jest.mock('./Export.messages', () => {
  return {
    export: {
      id: 'cboard.components.Settings.Export.export',
      defaultMessage: 'Export'
    },
    exportSecondary: {
      id: 'cboard.components.Settings.Export.exportSecondary',
      defaultMessage: 'Backup your boards'
    }
  };
});

const COMPONENT_PROPS = {
  onExportClick: () => {},
  onClose: () => {}
};

describe('Export tests', () => {
  test('default renderer', () => {
    shallowMatchSnapshot(<Export {...COMPONENT_PROPS} />);
  });

  test('menu behavior', () => {
    const wrapper = shallow(<Export {...COMPONENT_PROPS} />);
    let tree = toJson(wrapper);
    expect(tree).toMatchSnapshot();

    let exportMenu = wrapper.find('#export-menu').get(0);
    expect(exportMenu.props.open).toBe(false);

    const exportButton = wrapper.find('#export-button');
    exportButton.simulate('click', { currentTarget: 'someElement' });

    exportMenu = wrapper.find('#export-menu').get(0);
    expect(exportMenu.props.open).toBe(true);

    tree = toJson(wrapper);
    expect(tree).toMatchSnapshot();
  });

  test('loading behavior', () => {
    const wrapper = shallow(<Export {...COMPONENT_PROPS} />);
    let tree = toJson(wrapper);
    expect(tree).toMatchSnapshot();

    wrapper.instance().onExportClick = jest.fn(type => {
      wrapper.setState({
        loading: true,
        exportMenu: null
      });
    });

    let exportMenu = wrapper.find('#export-menu').get(0);
    expect(exportMenu.props.open).toBe(false);

    const exportButton = wrapper.find('#export-button');
    exportButton.simulate('click', { currentTarget: 'someElement' });

    exportMenu = wrapper.find('#export-menu').get(0);
    expect(exportMenu.props.open).toBe(true);

    tree = toJson(wrapper);
    expect(tree).toMatchSnapshot();

    const openboardOption = wrapper.find('WithStyles(MenuItem)').get(1);
    openboardOption.props.onClick();

    const spinnerWrapper = wrapper.find('.Export__ButtonContainer--spinner');
    expect(spinnerWrapper.length).toBe(1);

    expect(wrapper.find('#export-button').get(0).props.disabled).toBe(true);

    tree = toJson(wrapper);
    expect(tree).toMatchSnapshot();
  });
});
