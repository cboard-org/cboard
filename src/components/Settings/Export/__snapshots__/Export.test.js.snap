// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Export tests default renderer 1`] = `
<div
  className="Export"
>
  <WithStyles(FullScreenDialog)
    onClose={[Function]}
    open={true}
    title={
      <FormattedMessage
        defaultMessage="Export"
        id="cboard.components.Settings.Export.export"
        values={Object {}}
      />
    }
  >
    <WithStyles(Paper)>
      <WithStyles(List)>
        <WithStyles(ListItem)>
          <WithStyles(ListItemText)
            primary={
              <FormattedMessage
                defaultMessage="Export"
                id="cboard.components.Settings.Export.export"
                values={Object {}}
              />
            }
            secondary={
              <FormattedMessage
                defaultMessage="Backup your boards"
                id="cboard.components.Settings.Export.exportSecondary"
                values={
                  Object {
                    "link": <a
                      href="http://www.openboardformat.org/"
                    >
                      OpenBoard
                    </a>,
                  }
                }
              />
            }
          />
          <WithStyles(ListItemSecondaryAction)>
            <div
              className="Export__ButtonContainer"
            >
              <WithStyles(Button)
                disabled={false}
                id="export-button"
                onClick={[Function]}
              >
                <FormattedMessage
                  defaultMessage="Export"
                  id="cboard.components.Settings.Export.export"
                  values={Object {}}
                />
              </WithStyles(Button)>
              <WithStyles(Menu)
                anchorEl={null}
                id="export-menu"
                onClose={[Function]}
                open={false}
              >
                <WithStyles(MenuItem)
                  onClick={[Function]}
                >
                  Cboard
                </WithStyles(MenuItem)>
                <WithStyles(MenuItem)
                  onClick={[Function]}
                >
                  OpenBoard
                </WithStyles(MenuItem)>
                <WithStyles(MenuItem)
                  onClick={[Function]}
                >
                  PDF
                </WithStyles(MenuItem)>
              </WithStyles(Menu)>
            </div>
          </WithStyles(ListItemSecondaryAction)>
        </WithStyles(ListItem)>
      </WithStyles(List)>
    </WithStyles(Paper)>
  </WithStyles(FullScreenDialog)>
</div>
`;

exports[`Export tests loading behavior 1`] = `
<div
  className="Export"
>
  <WithStyles(FullScreenDialog)
    onClose={[Function]}
    open={true}
    title={
      <FormattedMessage
        defaultMessage="Export"
        id="cboard.components.Settings.Export.export"
        values={Object {}}
      />
    }
  >
    <WithStyles(Paper)>
      <WithStyles(List)>
        <WithStyles(ListItem)>
          <WithStyles(ListItemText)
            primary={
              <FormattedMessage
                defaultMessage="Export"
                id="cboard.components.Settings.Export.export"
                values={Object {}}
              />
            }
            secondary={
              <FormattedMessage
                defaultMessage="Backup your boards"
                id="cboard.components.Settings.Export.exportSecondary"
                values={
                  Object {
                    "link": <a
                      href="http://www.openboardformat.org/"
                    >
                      OpenBoard
                    </a>,
                  }
                }
              />
            }
          />
          <WithStyles(ListItemSecondaryAction)>
            <div
              className="Export__ButtonContainer"
            >
              <WithStyles(Button)
                disabled={false}
                id="export-button"
                onClick={[Function]}
              >
                <FormattedMessage
                  defaultMessage="Export"
                  id="cboard.components.Settings.Export.export"
                  values={Object {}}
                />
              </WithStyles(Button)>
              <WithStyles(Menu)
                anchorEl={null}
                id="export-menu"
                onClose={[Function]}
                open={false}
              >
                <WithStyles(MenuItem)
                  onClick={[Function]}
                >
                  Cboard
                </WithStyles(MenuItem)>
                <WithStyles(MenuItem)
                  onClick={[Function]}
                >
                  OpenBoard
                </WithStyles(MenuItem)>
                <WithStyles(MenuItem)
                  onClick={[Function]}
                >
                  PDF
                </WithStyles(MenuItem)>
              </WithStyles(Menu)>
            </div>
          </WithStyles(ListItemSecondaryAction)>
        </WithStyles(ListItem)>
      </WithStyles(List)>
    </WithStyles(Paper)>
  </WithStyles(FullScreenDialog)>
</div>
`;

exports[`Export tests loading behavior 2`] = `
<div
  className="Export"
>
  <WithStyles(FullScreenDialog)
    onClose={[Function]}
    open={true}
    title={
      <FormattedMessage
        defaultMessage="Export"
        id="cboard.components.Settings.Export.export"
        values={Object {}}
      />
    }
  >
    <WithStyles(Paper)>
      <WithStyles(List)>
        <WithStyles(ListItem)>
          <WithStyles(ListItemText)
            primary={
              <FormattedMessage
                defaultMessage="Export"
                id="cboard.components.Settings.Export.export"
                values={Object {}}
              />
            }
            secondary={
              <FormattedMessage
                defaultMessage="Backup your boards"
                id="cboard.components.Settings.Export.exportSecondary"
                values={
                  Object {
                    "link": <a
                      href="http://www.openboardformat.org/"
                    >
                      OpenBoard
                    </a>,
                  }
                }
              />
            }
          />
          <WithStyles(ListItemSecondaryAction)>
            <div
              className="Export__ButtonContainer"
            >
              <WithStyles(Button)
                disabled={false}
                id="export-button"
                onClick={[Function]}
              >
                <FormattedMessage
                  defaultMessage="Export"
                  id="cboard.components.Settings.Export.export"
                  values={Object {}}
                />
              </WithStyles(Button)>
              <WithStyles(Menu)
                anchorEl="someElement"
                id="export-menu"
                onClose={[Function]}
                open={true}
              >
                <WithStyles(MenuItem)
                  onClick={[Function]}
                >
                  Cboard
                </WithStyles(MenuItem)>
                <WithStyles(MenuItem)
                  onClick={[Function]}
                >
                  OpenBoard
                </WithStyles(MenuItem)>
                <WithStyles(MenuItem)
                  onClick={[Function]}
                >
                  PDF
                </WithStyles(MenuItem)>
              </WithStyles(Menu)>
            </div>
          </WithStyles(ListItemSecondaryAction)>
        </WithStyles(ListItem)>
      </WithStyles(List)>
    </WithStyles(Paper)>
  </WithStyles(FullScreenDialog)>
</div>
`;

exports[`Export tests loading behavior 3`] = `
<div
  className="Export"
>
  <WithStyles(FullScreenDialog)
    onClose={[Function]}
    open={true}
    title={
      <FormattedMessage
        defaultMessage="Export"
        id="cboard.components.Settings.Export.export"
        values={Object {}}
      />
    }
  >
    <WithStyles(Paper)>
      <WithStyles(List)>
        <WithStyles(ListItem)>
          <WithStyles(ListItemText)
            primary={
              <FormattedMessage
                defaultMessage="Export"
                id="cboard.components.Settings.Export.export"
                values={Object {}}
              />
            }
            secondary={
              <FormattedMessage
                defaultMessage="Backup your boards"
                id="cboard.components.Settings.Export.exportSecondary"
                values={
                  Object {
                    "link": <a
                      href="http://www.openboardformat.org/"
                    >
                      OpenBoard
                    </a>,
                  }
                }
              />
            }
          />
          <WithStyles(ListItemSecondaryAction)>
            <div
              className="Export__ButtonContainer"
            >
              <WithStyles(CircularProgress)
                className="Export__ButtonContainer--spinner"
                size={25}
                thickness={7}
              />
              <WithStyles(Button)
                disabled={true}
                id="export-button"
                onClick={[Function]}
              >
                <FormattedMessage
                  defaultMessage="Export"
                  id="cboard.components.Settings.Export.export"
                  values={Object {}}
                />
              </WithStyles(Button)>
              <WithStyles(Menu)
                anchorEl={null}
                id="export-menu"
                onClose={[Function]}
                open={false}
              >
                <WithStyles(MenuItem)
                  onClick={[Function]}
                >
                  Cboard
                </WithStyles(MenuItem)>
                <WithStyles(MenuItem)
                  onClick={[Function]}
                >
                  OpenBoard
                </WithStyles(MenuItem)>
                <WithStyles(MenuItem)
                  onClick={[Function]}
                >
                  PDF
                </WithStyles(MenuItem)>
              </WithStyles(Menu)>
            </div>
          </WithStyles(ListItemSecondaryAction)>
        </WithStyles(ListItem)>
      </WithStyles(List)>
    </WithStyles(Paper)>
  </WithStyles(FullScreenDialog)>
</div>
`;

exports[`Export tests menu behavior 1`] = `
<div
  className="Export"
>
  <WithStyles(FullScreenDialog)
    onClose={[Function]}
    open={true}
    title={
      <FormattedMessage
        defaultMessage="Export"
        id="cboard.components.Settings.Export.export"
        values={Object {}}
      />
    }
  >
    <WithStyles(Paper)>
      <WithStyles(List)>
        <WithStyles(ListItem)>
          <WithStyles(ListItemText)
            primary={
              <FormattedMessage
                defaultMessage="Export"
                id="cboard.components.Settings.Export.export"
                values={Object {}}
              />
            }
            secondary={
              <FormattedMessage
                defaultMessage="Backup your boards"
                id="cboard.components.Settings.Export.exportSecondary"
                values={
                  Object {
                    "link": <a
                      href="http://www.openboardformat.org/"
                    >
                      OpenBoard
                    </a>,
                  }
                }
              />
            }
          />
          <WithStyles(ListItemSecondaryAction)>
            <div
              className="Export__ButtonContainer"
            >
              <WithStyles(Button)
                disabled={false}
                id="export-button"
                onClick={[Function]}
              >
                <FormattedMessage
                  defaultMessage="Export"
                  id="cboard.components.Settings.Export.export"
                  values={Object {}}
                />
              </WithStyles(Button)>
              <WithStyles(Menu)
                anchorEl={null}
                id="export-menu"
                onClose={[Function]}
                open={false}
              >
                <WithStyles(MenuItem)
                  onClick={[Function]}
                >
                  Cboard
                </WithStyles(MenuItem)>
                <WithStyles(MenuItem)
                  onClick={[Function]}
                >
                  OpenBoard
                </WithStyles(MenuItem)>
                <WithStyles(MenuItem)
                  onClick={[Function]}
                >
                  PDF
                </WithStyles(MenuItem)>
              </WithStyles(Menu)>
            </div>
          </WithStyles(ListItemSecondaryAction)>
        </WithStyles(ListItem)>
      </WithStyles(List)>
    </WithStyles(Paper)>
  </WithStyles(FullScreenDialog)>
</div>
`;

exports[`Export tests menu behavior 2`] = `
<div
  className="Export"
>
  <WithStyles(FullScreenDialog)
    onClose={[Function]}
    open={true}
    title={
      <FormattedMessage
        defaultMessage="Export"
        id="cboard.components.Settings.Export.export"
        values={Object {}}
      />
    }
  >
    <WithStyles(Paper)>
      <WithStyles(List)>
        <WithStyles(ListItem)>
          <WithStyles(ListItemText)
            primary={
              <FormattedMessage
                defaultMessage="Export"
                id="cboard.components.Settings.Export.export"
                values={Object {}}
              />
            }
            secondary={
              <FormattedMessage
                defaultMessage="Backup your boards"
                id="cboard.components.Settings.Export.exportSecondary"
                values={
                  Object {
                    "link": <a
                      href="http://www.openboardformat.org/"
                    >
                      OpenBoard
                    </a>,
                  }
                }
              />
            }
          />
          <WithStyles(ListItemSecondaryAction)>
            <div
              className="Export__ButtonContainer"
            >
              <WithStyles(Button)
                disabled={false}
                id="export-button"
                onClick={[Function]}
              >
                <FormattedMessage
                  defaultMessage="Export"
                  id="cboard.components.Settings.Export.export"
                  values={Object {}}
                />
              </WithStyles(Button)>
              <WithStyles(Menu)
                anchorEl="someElement"
                id="export-menu"
                onClose={[Function]}
                open={true}
              >
                <WithStyles(MenuItem)
                  onClick={[Function]}
                >
                  Cboard
                </WithStyles(MenuItem)>
                <WithStyles(MenuItem)
                  onClick={[Function]}
                >
                  OpenBoard
                </WithStyles(MenuItem)>
                <WithStyles(MenuItem)
                  onClick={[Function]}
                >
                  PDF
                </WithStyles(MenuItem)>
              </WithStyles(Menu)>
            </div>
          </WithStyles(ListItemSecondaryAction)>
        </WithStyles(ListItem)>
      </WithStyles(List)>
    </WithStyles(Paper)>
  </WithStyles(FullScreenDialog)>
</div>
`;
